### PR TITLE
Move API token into the separate file.

### PR DESCRIPTION
**Pull request by dethoter**
Fix of #3748. BTW, it's not clear what to do with old config. 
Should I add a check for old config and try to remove [repository.token] field from it every time user add a new token?
Or should I just prefer to use a token field from a new config over the old one?

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks for the PR! Looking good to me! Some thoughts:

* Could you add some tests for this? Both that the old and the new file works
* I think we may want to not print out the token to the console?
* Could the crates.io documentation get updated?

**Comment by alexcrichton** [comment]
cc @matklad (just to make you aware)

**Comment by matklad** [comment]
@alexcrichton 

What are your thoughts on

>Should I add a check for old config and try to remove [repository.token] field from it every time user add a new token?
Or should I just prefer to use a token field from a new config over the old one?

I think we definitely should remove support for old token location some time in the future, because leaving security-related backwards comparability code paths is not the right thing.

**Comment by alexcrichton** [comment]
Oh right sorry forgot to mention that. Yeah let's issue a warning if `.cargo/config` contains a token on `cargo publish`, and then we'll advertise this change and eventually remove that configuration.

**Comment by dethoter** [comment]
Got it. I'll update the PR by the end of the week

**Comment by dethoter** [comment]
After diving into the code I found that it's a bit tricky to operate both tokens (from `~/.cargo/config` and from `~/.cargo/credentials`) due to `ConfigValue::merge()` implementation. It just skips duplicated values. For example, if we have 'token' key in the both old and new config files **the second one will be ignored** during the `Config::load_values()` invocation.

So I propose the next options:
1. Ignore the value from `~/.cargo/config` during the merge. Read `~/.cargo/config` during the `publish` in order to check if old configuration exists.
2. Tune `ConfigValue::merge()` somehow in order to keep both values during the loading.

**Comment by alexcrichton** [comment]
Perhaps the token could be handled specially? We don't necessarily want to load it from all possible config files, right? Just the home dir?

**Comment by dethoter** [comment]
@alexcrichton thank you for advice and sorry for a long delay. I've updated the code according to your recommendations. I'll introduce tests on the next week.

**Comment by alexcrichton** [comment]
Thanks @dethoter! One thing I just remembered (sorry for the runaround here) is that let's actually hold off on the warning for now. Users which use Cargo across multiple channels (stable/beta/nightly) would otherwise have no way to *not* get warnings, so let's wait for this support to hit stable and then we can enable warnings.

**Comment by matklad** [comment]
I am now reading https://github.com/rust-lang/rfcs/pull/2006 which proposes to add alternative registers to crates.io, which can be super useful in corporate scenarios. 

One of the issues there is that when there are multiple registers, you probably want to use different access tokens for them. 

I wonder if we can future proof this pull request for that sort of requirements changes. What if we use TOML for credentials, and store them like

```
[crates.io]
token = "DEADBEEF"
```

?

That way, we'll be able to add additional keys in the future easily.

**Comment by matklad** [comment]
cc https://github.com/rust-lang/cargo/issues/3365

**Comment by alexcrichton** [comment]
Seems reasonable to me!

**Comment by dethoter** [comment]
@alexcrichton no problem. I'll remove the code that warns.

@matklad good proposal! It should be easy to switch file to TOML one. But there are some questions about compatibility.

The old config supports just one registry, so users ask for token like  
```cfg.get_string("registry.token")```.

A new one consists from different tokens for different registries. For example I'll store them like:
```
registry.<host1>.token
registry.<host2>.token
```

How can I handle a backward compatibility?  Should a user now ask for a concrete token like:
```cfg.get_string("registry.host1.token")```? Or maybe we should have some default token?

**Comment by matklad** [comment]
>How can I handle a backward compatibility? 

Cargo as a library is not `1.0`, so we can, and do, break backwards compatibility. However I think in this PR we can keep the API exactly the same by returning the first token from the table. 

Then, in the PR implementing #3365 we can switch API to `cfg.get_string("registry.host1.token")`.

Or do you think than it'd be easier to add multiple hosts from the start here? I am not super familiar with implementation here :)

**Comment by dethoter** [comment]
> However I think in this PR we can keep the API exactly the same by returning the first token from the table.

I believe it's the best option for now. I'll submit an update in a few days.

**Comment by dethoter** [comment]
During the implementation, I decided that it'd be better to return the old token if cargo couldn't find a token for a host. Returning the first from the map is actually a bad idea because it's associated with another host. If you have any better idea I'll be happy to discuss it.

**Comment by matklad** [comment]
@alexcrichton anything else to do here?

File: tests/registry.rs [hunk 1]
```diff
@@ -581,7 +581,8 @@ fn dev_dependency_not_used() {
 fn login_with_no_cargo_dir() {
     let home = paths::home().join("new-home");
     t!(fs::create_dir(&home));
-    assert_that(cargo_process().arg("login").arg("foo").arg("-v"),
+    assert_that(cargo_process().arg("login").arg("--host").arg(registry().to_string())
```
Comment (alexcrichton) [inline tests/registry.rs]: To clarify, this isn't required now is it? (The `--host` argument)

**Comment by alexcrichton** [comment]
This looks pretty good to me! My only hesitation here is the handling of the key in `registry.$key.token`. There's a transformation that's applied here but I'm not sure it's one we'd want to stabilize on? Elsewhere what we've done is either `registry.crates-io.token` or `registry.$url.token`, perhaps we could do that here as well?

File: tests/registry.rs (hunk 1 above)
Comment (dethoter) [inline tests/registry.rs]: `--host` argument isn't required during the `login` invocation. But if it's omitted `login` assumes using `SourceId::crates_io` in order to get host and therefore updating registry from a network. It slows down the tests and requires a network connection, so I specify the fake host for the tests. (please take a look on `src/bin/login` diff for details)

**Comment by dethoter** [comment]
We can use `registry.crates-io.token` approach as good as `registry.$url.token`. But in the case of using the second one, the actual key will look like `registry."$url".token` because TOML treats dots as separator and URL contains the dots.

Please choose how we should store the host key.

**Comment by alexcrichton** [comment]
I'm personally still ok with 

```toml
token = "..."
```

without actually specifying the registry at all. I think there's a reasonable interpretation even in a multi-registry world of that TOML file, it's just a token for the "default registry" (e.g. crates.io) and you can specify others in the file as well eventually.

*If* we want to have a `[registry]` table then yes, I would expect quoted urls so it would look like:

```toml
[registry.crates-io]
token = "..."

# or ...

[registry.'https://github.com/rust-lang/crates.io-index']
token = "..."
```

I would personally just go for a bare `token = "..."` for simplicity, but @matklad do you think we need to handle the multi registry case now?

**Comment by matklad** [comment]
>I would personally just go for a bare token = "..." for simplicity, but @matklad do you think we need to handle the multi registry case now?

Hm, I think that for multiple registries `registry."$url".token` scheme is actually **simpler** to implement than the `default registry` one. Like, if user executes `cargo login --host default registry`, you'll need to write code to save the token under bare `token` and not `registry.$url`. 

And looks like current PR already closes  #3365, that is, it already implements all the support for several hosts.

**Comment by alexcrichton** [comment]
> you'll need to write code to save the token under bare token and not registry.$url.

Oh not necessarily, I figured that once we implement this then `token` is just a fallback just as `registry.token` is today (in `.cargo/config`)

**Comment by dethoter** [comment]
I agree with @matklad, it would be simpler to store all the tokens like `registry."$url".token`. Let's stop on this approach?

Yes, this PR already closes #3365.

**Comment by matklad** [comment]
>Let's stop on this approach?

:+1:

**Comment by alexcrichton** [comment]
Ok but to be clear we don't even know what the "url of a registry" is today. The url we currently use in lots of places, the index, seems clearly inferior to the actual url like "https://crates.io". 

What I'm worried about is that trying to preemptively support multiple registries when we don't actually know what it will look like will drive us into a corner which is very difficult to support in the future. Let's say we encode the index url today, but we'd rather "https://crates.io" tomorrow? We now have to support both urls! What if we decide to use bare identifiers for all registries instead of urls? 

I find `token = "..."` to be the least invasive in terms of forwards compatibility. Issues like #3365 are good for the long run but there's not really much use in fixing it one-off in a PR like this? If we fix it we're not really any closer to supporting custom registries...

**Comment by matklad** [comment]
>Ok but to be clear we don't even know what the "url of a registry" is today. The url we currently use in lots of places, the index, seems clearly inferior to the actual url like "https://crates.io".

Hm, an excellent point! So yeah, it makes sense to support a single token only in this PR, and implement multiple registers support later. Fixing #3365 is not really important now, because it affects only crates.io maintainers. 

Aslo, `credentials` file is not really meant to be edited/viewed by humans, so we can use whatever format we like, as long as it is extensible.


Unrelated, but we might want to change token location in the docs over here: https://github.com/rust-lang/cargo/blob/master/src/doc/crates-io.md#before-your-first-publish.

**Comment by dethoter** [comment]
@alexcrichton you're right, I'll remove support for multiple hosts from this PR.

**Comment by dethoter** [comment]
@alexcrichton @matklad seems like appveyor build is failed due to nmake issue. Could you please restart the build or point me how can I fix it?

**Comment by alexcrichton** [comment]
Ah no worries that should hopefully be fixed by the time this reaches the head of the queue. Thanks again @dethoter!

@bors: r+

**Comment by dethoter** [comment]
@alexcrichton @matklad  thank you for supporting.

**Comment by jan-hudec** [comment]
@dethoter, the `cargo::util::config::Config::load_config` method looks for `.cargo/config` in current working directory and all its parents. Why does the `load_credentials` method not do the same except looking for `.cargo/credentials`?

**Comment by dethoter** [comment]
Because `.cargo/credentials` is global, but `.cargo/config` could be both global and local.

**Comment by jan-hudec** [comment]
@dethoter, but why can't credentials be local too? If the *local* `.cargo/config` overrides registry index/host, the corresponding credentials need to be stored somewhere. And local `.cargo/credentials` is the most logical (and without named registries only) place for it.

**Comment by dethoter** [comment]
AFAIK cargo will not support unnamed registries. So there is no reason to handle credentials per project. `$CARGO_HOME/credentials` is the only place to store tokens/passwords for all the registries. Am I right, @alexcrichton?

**Comment by alexcrichton** [comment]
The precise design for multiple registries is [still up in the air](https://github.com/rust-lang/rfcs/pull/2141), but I think it's reasonable to consider loading any `.cargo/credentials` a possibility. The format will probably get tweaked (extended) slightly to support custom registries as well.

**Pull request merged**